api/v1alpha1/stage_status_types.go

```diff
@@ -0,0 +1,79 @@
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// StageStatusPhase defines the observed stage phase
+// +kubebuilder:validation:Enum={Progressing,Succeeded,Failed}
+type StageStatusPhase string
+
+const (
+	PhaseProgressing StageStatusPhase = "Progressing"
+	PhaseSucceeded   StageStatusPhase = "Succeeded"
+	PhaseFailed      StageStatusPhase = "Failed"
+)
+
+// StageStatus defines the observed stage status
+type StageStatus struct {
+	Name       string           `json:"name"`
+	Phase      StageStatusPhase `json:"stage,omitempty"`
+	Message    string           `json:"message,omitempty"`
+	StartedAt  *metav1.Time     `json:"startedAt,omitempty"`
+	FinishedAt *metav1.Time     `json:"finishedAt,omitempty"`
+}
+
+// GetStageStatus returns a pointer to the Status.Stages slice
+func (in *ProgressiveRollout) GetStageStatus() *[]StageStatus {
+	return &in.Status.Stages
+}
+
+// NewStageStatus adds a new StageStatus
+func (in *ProgressiveRollout) NewStageStatus(name, message string, phase StageStatusPhase) StageStatus {
+	return StageStatus{
+		Name:       name,
+		Phase:      phase,
+		Message:    message,
+		StartedAt:  nil,
+		FinishedAt: nil,
+	}
+}
+
+// SetStageStatus sets the corresponding StageStatus in stageStatus to newStatus
+// - If a stage doesn't exist, it will be added to StageStatus slice
+// - If a stage already exists it will be updated
+func SetStageStatus(stageStatus *[]StageStatus, newStatus StageStatus, updateTime *metav1.Time) {
```

Comment:
this should be a slice, not a pointer to a slice
```suggestion
func SetStageStatus(stageStatus []StageStatus, newStatus StageStatus, updateTime *metav1.Time) {
```

Reply:
on second read, i see what you've done here. I'm not sure this function signature is really the best here.
unlike `NewStageStatus`, that one should prob be a method of `ProgressiveRollout`.

Something like:
```suggestion
func (p *ProgressiveRollout) SetStageStatus(newStatus StageStatus, updateTime *metav1.Time) {
```

The name `SetStageStatus` i struggled with a little bit... , but that's secondary.

Reply:
on third read :-) i'd prob suggest to create two functions instead, like `AddStageStatus` and `UpdateStageStatus`, and handle the `FindStageStatus` in the `updateStageStatus` at line 351

Reply:
Did the first two suggested changes - but opted not to make the change to the method just now as it provides a simple single interface to make these change. Perhaps we can rename it to make it clearer?